### PR TITLE
Refactor and update existing ml roles

### PR DESCRIPTION
**Pull request by b4sjoo**
### Description
This PR refactor and update exisiting ml roles

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: config/roles.yml
Comment (lezzago) [inline config/roles.yml:278]: Can we add this permission here: `cluster:admin/opensearch/ml/config/get`
It is needed to fetch the agent id. Also will the `cluster:admin/opensearch/ml/execute` provide the ability to execute agents?

Comment (jngz-es) [reply config/roles.yml:278]: We think the purpose of adding this role allows user-admin to grand an end-user to have permissions to use query assistant, not let end-user to create a new query assistant. We don't think the agent related permission needed.
```
cluster:admin/opensearch/ml/agents/register
cluster:admin/opensearch/ml/agents/get
cluster:admin/opensearch/ml/agents/search
cluster:admin/opensearch/ml/agents/delete
```
@b4sjoo , did you verify with this role, a user can use query assistant?

Comment (lezzago) [reply config/roles.yml:278]: I am also not sure if we need these permissions as there shouldn't be any memory stored for each natural language to query request.
```
cluster:admin/opensearch/ml/memory/conversation/create
cluster:admin/opensearch/ml/memory/interaction/create
```

I believe only the execute agent and get config apis are used for the query assistant feature: https://github.com/opensearch-project/dashboards-observability/blob/2.x/server/routes/query_assist/utils/agents.ts

File: config/roles.yml
Comment (ylwu-amzn) [inline config/roles.yml:290]: Don't forget to configure in security dashboard plugin https://github.com/opensearch-project/security-dashboards-plugin/pull/1182/files

File: config/roles.yml
Comment (ylwu-amzn) [inline config/roles.yml:284]: Will it cause any BWC issue if user upgrade to 2.13 from an old cluster?

File: config/roles.yml
Comment (b4sjoo) [reply config/roles.yml:290]: See https://github.com/opensearch-project/security-dashboards-plugin/pull/1838

File: config/roles.yml
Comment (b4sjoo) [reply config/roles.yml:284]: Will double check this with security team

**Comment by cwperks** [comment]
@b4sjoo @ylwu-amzn @lezzago 

If this PR and associated PR in [security-dashboards-plugin](https://github.com/opensearch-project/security-dashboards-plugin/pull/1838) are intended for 2.13 please mark this approved to signal to the security maintainers that it is good to go and also get signoff from @bbarani for inclusion in the release as this is post code freeze.

File: config/roles.yml
Comment (cwperks) [reply config/roles.yml:284]: It would cause bwc issues if it were a change in the role name. It is ok to change the permissions in a role. Cluster admins can either use securityadmin or a [new security api being introduced in 2.13](https://github.com/opensearch-project/documentation-website/pull/6634) to source changes to default roles into the security index.

Comment (cwperks) [reply config/roles.yml:284]: FYI changes to the role name are problematic because the role name is used in [`roles_mappings.yml`](https://opensearch.org/docs/latest/security/configuration/yaml/#roles_mappingyml) to map users to a role. If the role name changes, but the corresponding roles_mapping entry does not change, then those users would effectively lose access to functionality that they were able to previously use.

Comment (cwperks) [reply config/roles.yml:284]: Similarly, some plugins utilize roles injection for scheduled jobs to do permissions check on actions for scheduled jobs. If a role name changes, there is a risk that there are orphaned references to roles in disparate system indexes across the plugin ecosystem.

Comment (b4sjoo) [reply config/roles.yml:287]: Thanks for answering!

File: config/roles.yml
Comment (b4sjoo) [reply config/roles.yml:278]: Minimal set was found through e2e test on test os domain:
```
    "cluster:admin/opensearch/ml/execute",
    "cluster:admin/opensearch/ml/memory/conversation/create",
    "cluster:admin/opensearch/ml/memory/interaction/create",
    "cluster:admin/opensearch/ml/memory/interaction/update",
    "cluster:admin/opensearch/ml/predict",
    "cluster:monitor/state",
    "cluster:monitor/health"
```

File: config/roles.yml
Comment (peternied) [inline config/roles.yml:275]: The rename of this item can create some confusion.  For new clusters they will have `query_assistant_access`, for older clusters, they will not get this new name/permissions.  So you can have clusters running 2.13 that have either `ml_query_assistant_access` or `query_assistant_access`, even with the new upgrade API [1] admins might have trouble knowing when to use one or the other.

I'd recommend keeping the old name, if the new name is important, duplicate the role with both names to avoid misalignment.  Changing from `ml_query_assistant_access` -> `query_assistant_access` is a manual change for admins.

- [1] https://github.com/opensearch-project/documentation-website/pull/6634

**Review by peternied — CHANGES_REQUESTED** [review]

File: config/roles.yml
Comment (b4sjoo) [reply config/roles.yml:275]: Hi Peter, but this role has not been released yet, will that matter?

**Review by lezzago — APPROVED** [review]

**Comment by b4sjoo** [comment]
@peternied @cwperks Hi security team, just want to let you know we are good to ship both this PR and the [dashboard](https://github.com/opensearch-project/security-dashboards-plugin/pull/1838) one, I am going to notify @bbarani.  In the meantime, you can either merge this one first and wait @bbarani to sign-off the 2.13 backport, or wait until his sign-off to merge it to mainline and backport.

**Review by peternied — APPROVED** [review]

**Review by cwperks — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

**Comment by cwperks** [comment]
FYI This has already been [merged to 2.13](https://github.com/opensearch-project/security/pull/4157), this will get merged to main and 2.x shortly but the merge to 2.13 was done first.

**Pull request merged**
